how-to-update-to-php74-typed-property-without-bc-break/post.md

```diff
@@ -0,0 +1,171 @@
+How to Update to PHP 7.4 Typed Property Without BC Break with Rector
+====================================================================
+
+[Typed Property](https://wiki.php.net/rfc/typed_properties_v2) is one of the PHP 7.4 feature that allow to write that previously like this:
+
+```php
+namespace Lib;
+
+class SomeClass
+{
+    /** @var int */
+    public $a;
+
+    /** @var string */
+    protected $b;
+
+    /** @var bool */
+    private $c;
+}
+```
+
+to this:
+
+```php
+namespace Lib;
+
+class SomeClass
+{
+    public int $a;
+
+    protected string $b;
+
+    private bool $c;
+}
+```
+
+If you follow [Semver](https://semver.org) for versioning, and you don't want to update to major change, eg: version 1 to version 2, changing this will make Break Backward Compatibility, for example:
+
+```php
+namespace Lib;
+
+class SomeClass
+{
+    protected string $b;
+}
+```
+
+has child in application consumer:
+
+```php
+namespace App;
+
+use Lib\SomeClass;
+
+class AChild extends SomeClass
+{
+    protected $b;
+}
+```
+
+will result a fatal error:
+
+```
+Fatal error: Type of AChild::$b must be string (as in class SomeClass)
+```
+
+see https://3v4l.org/X9Yvd . To avoid that, you should only change to private modifier only, so, the change will only to private property:
+
+```diff
+namespace Lib;
+
+class SomeClass
+{
+    /** @var int */
+    public $a;
+
+    /** @var string */
+    protected $b;
+
+-    /** @var bool */
+-    private $c;
++    private bool $c;
+}
+```
+
+Want to automate that? You can use [Rector](https://github.com/rectorphp/rector) for it. First, let say, we have a library with the following structure:
+
+```bash
+lib
+├── composer.json
+├── composer.lock
+├── src
+│   └── SomeClass.php
+```
+
+with composer.json config like this:
+
+```javascript
+{
+    "require": {
+        "php": "^7.4"
+    },
+    "autoload": {
+        "psr-4": {
+            "Lib\\": "src/"
+        }
+    }
+}
+
+```
+
+Now, what you need is require the rector as dev dependency:
+
+```bash
+cd lib/
+composer require --dev rector/rector
+```
+
+Then, create a `rector.php` configuration inside the `lib`:
```

Comment:
I wonder, why inside `lib` directory?
The `rector.php` config should be always just 1 and located inside the root directory, to avoid any path confuctions.

Reply:
What I meant about lib is a package that redistributable, inside lib is a root of lib directory, should I update wording about it?

Reply:
I updated wording about usage "lib" as redistributable package that can be consumed in applications so place `rector.php` in the root of `lib` dir https://github.com/samsonasik/rector-blogposts/pull/2/commits/47ef866dbd1ac2f392b40911bd9cb8ff277d5210

Reply:
Ah, I see. I think the location is always root by default. If you have package or project, the composer and dependency structure is the same.

That's why I was confused about extra directory :)